File: android/app/src/main/java/com/mattermost/share/ShareUtils.java
package com.mattermost.share;

import android.app.Activity;
import android.content.Context;
import android.graphics.Bitmap;
import android.graphics.BitmapFactory;
import android.media.MediaMetadataRetriever;
import android.net.Uri;
import android.webkit.URLUtil;

import com.facebook.react.bridge.Arguments;
import com.facebook.react.bridge.ReadableMap;
import com.facebook.react.bridge.WritableMap;
import com.mattermost.helpers.RealPathUtil;

import java.io.File;
import java.io.FileOutputStream;
import java.io.OutputStream;
import java.io.UnsupportedEncodingException;
import java.net.URLDecoder;
import java.util.UUID;

public class ShareUtils {
    public static ReadableMap getTextItem(String text) {
        WritableMap map = Arguments.createMap();
        map.putString("value", text);
        map.putString("type", "");
        map.putBoolean("isString", true);
        return map;
    }

    public static ReadableMap getFileItem(Activity activity, Uri uri) {
        WritableMap map = Arguments.createMap();
        String filePath = RealPathUtil.getRealPathFromURI(activity, uri);
        if (filePath == null) {
            return null;
        }

        File file = new File(filePath);
        String type = RealPathUtil.getMimeTypeFromUri(activity, uri);
        if (type != null) {
            if (type.startsWith("image/")) {
                BitmapFactory.Options bitMapOption = ShareUtils.getImageDimensions(filePath);
                map.putInt("height", bitMapOption.outHeight);
                map.putInt("width", bitMapOption.outWidth);

            } else if (type.startsWith("video/")) {
                File cacheDir = new File(activity.getCacheDir(), RealPathUtil.CACHE_DIR_NAME);
                addVideoThumbnailToMap(cacheDir, activity.getApplicationContext(), map, "file://" + filePath);
            }
        } else {
            type = "application/octet-stream";
        }

        map.putString("value", "file://" + filePath);
        map.putDouble("size", (double) file.length());
        map.putString("filename", file.getName());
        map.putString("type", type);
        map.putString("extension", RealPathUtil.getExtension(filePath).replaceFirst(".", ""));
        map.putBoolean("isString", false);
        return map;
    }

    public static BitmapFactory.Options getImageDimensions(String filePath) {
        BitmapFactory.Options bitMapOption = new BitmapFactory.Options();
        bitMapOption.inJustDecodeBounds=true;
        BitmapFactory.decodeFile(filePath, bitMapOption);
        return bitMapOption;
    }

    private static void addVideoThumbnailToMap(File cacheDir, Context context, WritableMap map, String filePath) {
        String fileName = ("thumb-" + UUID.randomUUID().toString()) + ".png";
        OutputStream fOut = null;

        try {
            File file = new File(cacheDir, fileName);
            Bitmap image = getBitmapAtTime(context, filePath, 1);
            if (file.createNewFile()) {
                fOut = new FileOutputStream(file);
                image.compress(Bitmap.CompressFormat.PNG, 100, fOut);
                fOut.flush();
                fOut.close();

                map.putString("videoThumb", "file://" + file.getAbsolutePath());
                map.putInt("width", image.getWidth());
                map.putInt("height", image.getHeight());
            }
        } catch (Exception ignored) {
        }
    }

    private static Bitmap getBitmapAtTime(Context context, String filePath, int time) {
        try {
            MediaMetadataRetriever retriever = new MediaMetadataRetriever();
            if (URLUtil.isFileUrl(filePath)) {
                String decodedPath;
                try {
                    decodedPath = URLDecoder.decode(filePath, "UTF-8");
                } catch (UnsupportedEncodingException e) {
                    decodedPath = filePath;
                }

                retriever.setDataSource(decodedPath.replace("file://", ""));
            } else if (filePath.contains("content://")) {
                retriever.setDataSource(context, Uri.parse(filePath));
            }

            Bitmap image = retriever.getFrameAtTime(time * 1000, MediaMetadataRetriever.OPTION_CLOSEST_SYNC);
            retriever.release();
            return image;
        } catch (Exception e) {
            throw new IllegalStateException("File doesn't exist or not supported");
        }
    }
}

File: libraries/@mattermost/rnshare/android/src/main/java/com/mattermost/rnshare/ShareUtils.kt
package com.mattermost.rnshare

import android.app.Activity
import android.content.Context
import android.graphics.Bitmap
import android.graphics.BitmapFactory
import android.media.MediaMetadataRetriever
import android.net.Uri
import android.webkit.URLUtil
import androidx.core.net.toUri
import com.facebook.react.bridge.Arguments
import com.facebook.react.bridge.ReadableMap
import com.facebook.react.bridge.WritableMap
import com.mattermost.rnshare.helpers.RealPathUtil
import java.io.File
import java.io.FileOutputStream
import java.io.OutputStream
import java.io.UnsupportedEncodingException
import java.net.URLDecoder
import java.util.UUID

object ShareUtils {
    fun getTextItem(text: String?): ReadableMap {
        val map = Arguments.createMap()
        map.putString("value", text)
        map.putString("type", "")
        map.putBoolean("isString", true)
        return map
    }

    fun getFileItem(activity: Activity, uri: Uri): ReadableMap? {
        val map = Arguments.createMap()
        val filePath: String = RealPathUtil.getRealPathFromURI(activity, uri)
                ?: return null

        val file = File(filePath)
        var type = RealPathUtil.getMimeTypeFromUri(activity, uri)
        if (type?.startsWith("image/") == true) {
            val bitMapOption = getImageDimensions(filePath)
            map.putInt("height", bitMapOption.outHeight)
            map.putInt("width", bitMapOption.outWidth)
        } else if (type?.startsWith("video/") == true) {
            val cacheDir = File(activity.cacheDir, RealPathUtil.CACHE_DIR_NAME)
            addVideoThumbnailToMap(cacheDir, activity.applicationContext, map, "file://$filePath")
        } else {
            type = "application/octet-stream"
        }

        map.putString("value", "file://$filePath")
        map.putDouble("size", file.length().toDouble())
        map.putString("filename", file.name)
        map.putString("type", type)
        map.putString("extension", RealPathUtil.getExtension(filePath)?.replaceFirst(".".toRegex(), "") ?: "")
        map.putBoolean("isString", false)
        return map
    }

    private fun getImageDimensions(filePath: String): BitmapFactory.Options {
        val bitMapOption = BitmapFactory.Options()
        bitMapOption.inJustDecodeBounds = true
        BitmapFactory.decodeFile(filePath, bitMapOption)
        return bitMapOption
    }

    private fun addVideoThumbnailToMap(cacheDir: File, context: Context, map: WritableMap, filePath: String) {
        val fileName = ("thumb-" + UUID.randomUUID().toString()) + ".png"
        val fOut: OutputStream?

        try {
            val file = File(cacheDir, fileName)
            val image = getBitmapAtTime(context, filePath)
            if (file.createNewFile()) {
                fOut = FileOutputStream(file)
                image!!.compress(Bitmap.CompressFormat.PNG, 100, fOut)
                fOut.flush()
                fOut.close()

                map.putString("videoThumb", "file://" + file.absolutePath)
                map.putInt("width", image.width)
                map.putInt("height", image.height)
            }
        } catch (ignored: Exception) {
        }
    }

    private fun getBitmapAtTime(context: Context, filePath: String): Bitmap? {
        try {
            val time = 1
            val retriever = MediaMetadataRetriever()
            if (URLUtil.isFileUrl(filePath)) {
                val decodedPath = try {
                    URLDecoder.decode(filePath, "UTF-8")
                } catch (e: UnsupportedEncodingException) {
                    filePath
                }

                retriever.setDataSource(decodedPath.replace("file://", ""))
            } else if (filePath.contains("content://")) {
                retriever.setDataSource(context, filePath.toUri())
            }

            val image = retriever.getFrameAtTime((time * 1000).toLong(), MediaMetadataRetriever.OPTION_CLOSEST_SYNC)
            retriever.release()
            return image
        } catch (e: Exception) {
            throw IllegalStateException("File doesn't exist or not supported")
        }
    }
}
